### PR TITLE
Second patch of bumping adapters branch

### PR DESCRIPTION
**Pull request by omarahmed1111**
bump adapters branch to catch to main till this [commit](https://github.com/oneapi-src/unified-runtime/commit/b12b606c35bb7469c1dfdf932e2917238635f02d). This is a continuation of this [PR](https://github.com/oneapi-src/unified-runtime/pull/1016)

Intel/LLVM testing and adding some fixes on intel/llvm side associated to this changes: [here](https://github.com/intel/llvm/pull/11771)

### CODE REVIEW

**Comment by omarahmed1111** [comment]
Hi @pbalcer , the changes here were failing in python-basic test, the reason of the failure is that `ur.py` file was not able to catch the function pointer for `cooperative kernels` experimental feature from the loader. This is also happening for adding any experimental feature in this branch. I also tried with the current adapters branch and the same issue was there. Do you have any idea what could have caused that? or where does `ur.py` look into the loader (files linked to generate `libur_loader.so`) when it tries to link the ur loader entries with the python part? Thanks!

**Comment by pbalcer** [comment]
> Hi @pbalcer , the changes here were failing in python-basic test, the reason of the failure is that `ur.py` file was not able to catch the function pointer for `cooperative kernels` experimental feature from the loader. This is also happening for adding any experimental feature in this branch. I also tried with the current adapters branch and the same issue was there. Do you have any idea what could have caused that? or where does `ur.py` look into the loader (files linked to generate `libur_loader.so`) when it tries to link the ur loader entries with the python part? Thanks!

I'll take a look tomorrow, I'm guessing experimental features just aren't getting pulled into the python api properly.

But I think a better question is: do we need the python bindings in the first place? they are annoying to keep in a working state, we barely test them, and, as far as I know, no one uses them. They are a leftover from the origin of the project as a L0 loader fork. @kbenzie what do you think?

**Comment by kbenzie** [comment]
> But I think a better question is: do we need the python bindings in the first place? they are annoying to keep in a working state, we barely test them, and, as far as I know, no one uses them. They are a leftover from the origin of the project as a L0 loader fork. @kbenzie what do you think?

This is a good question. We don't have a use case for these right now. I'm not attached to the Python bindings at all, and its been noted in #136 and #135 that they are not idiomatic either.

Practically, removing the Python bindings while synchronizing the main and adapters branches might be a bit painful.

**Comment by pbalcer** [comment]
> not able to catch the function pointer for `cooperative kernels` experimental feature from the loader. 

#1036 

> Practically, removing the Python bindings while synchronizing the main and adapters branches might be a bit painful.

I was more thinking about just suppressing the python test errors for the time being. But let's revisit the topic once all this branch merging is sorted.

**Comment by kbenzie** [comment]
> I was more thinking about just suppressing the python test errors for the time being. But let's revisit the topic once all this branch merging is sorted.

Ah yes, disabling the basic python test would be simpler. Okay @omarahmed1111 lets to that to unblock this, I think commenting out the `add_subdirectory(python)` in `test/CMakeLists.txt` would be the easiest way.

**Review by EwanC — APPROVED** [review]
command-buffer changes LGTM, this PR is very helpful in unblocking some other changes we have pending.

**Review by isaacault — APPROVED** [review]
Bindless Images LGTM!

File: source/adapters/cuda/common.cpp
Comment (fabiomestre) [inline source/adapters/cuda/common.cpp:114]: Why was this change necessary?

Comment (omarahmed1111) [reply source/adapters/cuda/common.cpp:114]: `cerr` was catching a weird seg fault that was avoided with `fprintf`

Comment (fabiomestre) [reply source/adapters/cuda/common.cpp:114]: It worries me a bit that we never found out what exactly was causing the segfault. It might be a segfault / undefined behavior in some other part of the adapter that for some reason tends to crash inside std::cerr. I think we should investigate this further. Maybe running a sanitizer or valgrind might help.

However, since it seems to pass with the change to fprintf, maybe this PR can still be merged as long as we look into it afterwards.

Comment (kbenzie) [reply source/adapters/cuda/common.cpp]: > However, since it seems to pass with the change to fprintf, maybe this PR can still be merged as long as we look into it afterwards.

Agreed, open an issue to investigate this later.

Comment (omarahmed1111) [reply source/adapters/cuda/common.cpp]: The issue here was that the changes from this [PR](https://github.com/oneapi-src/unified-runtime/pull/1016) was not included due to rebasing on adapters revert commit, so the deepbind loading of the loaders was still there, and that was causing this issue. once I have included the revert of the revert commit (the commit for this [PR](https://github.com/oneapi-src/unified-runtime/pull/1016)) and the deepbind removal was included. This problem have been fixed. So, it should be fixed now.

**Review by Seanst98 — CHANGES_REQUESTED** [review]
My commit adding unique addressing modes per dimension for the bindless images extension found [here](https://github.com/oneapi-src/unified-runtime/pull/1014/commits/5640ea63f5995a0636ff1f264a8c17b73a26cfef) and [PR here](https://github.com/oneapi-src/unified-runtime/pull/863) has been added in this bump. 

At the time, UR did not support C style arrays so the above commit compromised by using 3 different vars to represent each dimension and was merged into main. A new commit was pushed to main and merged [here](https://github.com/oneapi-src/unified-runtime/pull/870/commits/7217fd7a540d22a6eee153f15bffea2beceea725) and [PR here](https://github.com/oneapi-src/unified-runtime/pull/870) which added C style array support and modified the extension to use C style arrays. 

I can't find the commits which update to using C style arrays in this bump. Can we add it?

**Comment by kbenzie** [comment]
> I can't find the commits which update to using C style arrays in this bump. Can we add it?

We've split up the main bumps because that was easier to handle the issues which arose from this process. We will be pulling in the rest of the changes from main in a third PR. Then, I think we'll merge them all back to back so they should land in adapters branch at the same time.

**Review by Seanst98 — APPROVED** [review]

**Review by fabiomestre — APPROVED** [review]

**Comment by omarahmed1111** [comment]
Due to a windows failure that will be fixed [here](https://github.com/oneapi-src/unified-runtime/pull/1084) and to make it easier, we will combine the second and third bump into one [here](https://github.com/oneapi-src/unified-runtime/pull/1072). So closing this for now.

**Pull request closed without merging**
